### PR TITLE
Correct exponent_max to conform to IEEE definition

### PR DESCRIPTION
**Pull request by musm**
rebased from https://github.com/JuliaLang/julia/pull/24978

cc @JeffreySarnoff

### CODE REVIEW

File: base/float.jl [hunk 1]
```diff
@@ -78,11 +78,35 @@ for T in (Float16, Float32, Float64)
...
-    @eval exponent_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)) - exponent_bias(T))
+    @eval exponent_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)) - exponent_bias(T) - 1)
     # maximum float exponent without bias
     @eval exponent_raw_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)))
 end
 
+"""
+    exponent_max(T)
+s
+Maximum exponent a real floating point number of type `T` can have.
```
Comment (musm) [inline base/float.jl:89]:
```suggestion

Maximum [`exponent`](@ref) value for a real floating point number of type `T`.
```

File: base/float.jl [hunk 2]
```diff
@@ -78,11 +78,35 @@ for T in (Float16, Float32, Float64)
...
+The value `exponent_max(T) + 1` is a possible value of the exponent field
+with bias, which might be used as sentinel value for `Inf` and `NaN`.
+"""
+function exponent_max end
+
+"""
+    exponent_raw_max(T)
+
+The maximum value of the exponent field of floating point type `T` without bias,
+i.e. the maximum integer value representable by `exponent_bits(T)` bits.
```
Comment (musm) [inline base/float.jl:106]:
```suggestion
Maximum value of the [`exponent`](@ref) field for a floating point number of type `T` without bias,
i.e. the maximum integer value representable by [`exponent_bits(T)`](@ref) bits.
```

File: base/float.jl [hunk 3]
```diff
@@ -78,11 +78,35 @@ for T in (Float16, Float32, Float64)
...
 end
 
+"""
+    exponent_max(T)
+
+Maximum [`exponent`](@ref) value for a floating point number of type `T`.
+
+# Examples
+```jldoctest
+julia> exponent_max(Float64)
```
Comment (musm) [inline base/float.jl:93]:
```suggestion
julia> Base.exponent_max(Float64)
```

File: base/float.jl [hunk 4]
```diff
@@ -78,11 +78,35 @@ for T in (Float16, Float32, Float64)
...
     @eval exponent_bias(::Type{$T}) = $(Int(exponent_one(T) >> significand_bits(T)))
     # maximum float exponent
-    @eval exponent_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)) - exponent_bias(T))
+    @eval exponent_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)) - exponent_bias(T) - 1)
     # maximum float exponent without bias
     @eval exponent_raw_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)))
 end
 
+"""
+    exponent_max(T)
```
Comment (musm) [inline base/float.jl:87]: @fredrikekre  I hesitate to add these docstring (rebased and modified from the previous PR) it was suggested to add for clarification. My concern: we don't export this method, so does documenting the method raise an issue? I know we prefix the module name for public API  that are  *not* exported but still part of the formal API. In this instance, I don't recall what the policy is, but it does seem helpful to include some documentation even if they are primarily internal methods. Should I convert them to comments instead, perhaps?

Comment (Keno) [reply base/float.jl:87]: We document internal functions all over the place. More documentation is always better.

**Comment by mschauer** [comment]
Looks good, also https://juliahub.com reports no additional use.

File: base/float.jl (hunk 4 above)
Comment (mschauer) [reply base/float.jl:87]: Yes, this was deliberate

File: base/float.jl [hunk 5]
```diff
@@ -78,11 +78,35 @@ for T in (Float16, Float32, Float64)
...
+
+Maximum [`exponent`](@ref) value for a floating point number of type `T`.
+
+# Examples
+```jldoctest
+julia> Base.exponent_max(Float64)
+1023
+```
+
+The value `exponent_max(T) + 1` is a possible value of the exponent field
```
Comment (musm) [inline base/float.jl:97]:
```suggestion
Note, `exponent_max(T) + 1` is a possible value of the exponent field
```

File: base/float.jl [hunk 6]
```diff
@@ -78,11 +78,35 @@ for T in (Float16, Float32, Float64)
...
+with bias, which might be used as sentinel value for `Inf` and `NaN`.
```
Comment (musm) [inline base/float.jl:98]:
```suggestion
with bias, which might be used as sentinel value for `Inf` or `NaN`.
```

**Comment by musm** [comment]
will merge within 48 hours. Original PR was approved by `@simonbyrne`

**Comment by goualard-f** [comment]
I do not get it. In Julia 1.6.0, Base.exponent_max still returns the "wrong value", e.g. :

julia> Base.exponent_max(Float64)
1024

**Comment by simeonschaub** [comment]
This was merged after 1.6 was branched off. But perhaps this change should be backported though?

**Comment by goualard-f** [comment]
I needed the correct definition but suffices to know for now that it will eventually make its way to the next release. In the meantime, I can rely on some workaround. Thanks.

**Comment by KristofferC** [comment]
This caused https://github.com/FedericoStra/ExactConversions.jl to start failing. Haven't looked into why, they probably just relied on the old behavior for some reason.

**Comment by JeffreySarnoff** [comment]
There are 3 lines in  https://github.com/FedericoStra/ExactConversions.jl that use `Base.exponent_max`.
The first line in both versions of `function maxcommon` and the first line of `function mincommon`.
The current source uses `(Base.exponent_max(T)-1)` which corrects for the old nonconformant `exponent_max`.
By removing the `-1` from each of the three expressions, their code ought not error with this PR.

**Comment by KristofferC** [comment]
Thanks for looking into it @JeffreySarnoff

**Pull request merged**